File: src/wire/frames.ts
// Wire protocol frames (ADR-0001 §single-ordered-stream, ADR-0002 §2).
// Shared by the Durable Object server and the browser client.
//
// One ordered stream per DO. The client tracks a single cursor (appliedSeq);
// `seq` is opaque (a stringified bigint). Confirmation rides on the same stream
// as data — `committed`/`rejected` correlate to a client `txId`; there is no
// second ack channel (ADR-0002 C1).
//
// Predicate/order fields are typed `unknown` here and tightened to the
// @tanstack/db expression IR in M5.

export type Cursor = string
export type TxId = string
export type RowOp = "insert" | "update" | "delete"

/** Placeholder for the @tanstack/db BasicExpression IR (tightened in M5). */
export type WireExpression = unknown
export type WireOrderBy = unknown

/** One row operation inside a `mut` frame. `cols` is a getChanges() diff; for
 *  an insert it is the full row (ADR-0001 D19). */
export interface MutOp {
  type: RowOp
  key: unknown
  cols?: Record<string, unknown>
}

export type ClientFrame =
  | {
      t: "sub"
      subId: string
      collection: string
      where?: WireExpression
      orderBy?: WireOrderBy
      limit?: number
      offset?: number
      cursor?: WireExpression
      since?: Cursor
    }
  | { t: "unsub"; subId: string }
  | { t: "mut"; txId: TxId; collection: string; ops: Array<MutOp> }
  | { t: "call"; txId: TxId; name: string; args: unknown }
  // One-shot paginated page fetch — a subset snapshot with NO live
  // subscription. Used for cursor load-more (scroll-back); deltas for the
  // window already flow via the live `sub` on the query's `where`.
  //
  // The frame is a serialized subset of @tanstack/db's `LoadSubsetOptions`
  // (ADR-0005): `where` is the base filter and `cursor` is its `CursorExpressions`,
  // carried RAW (whereFrom/whereCurrent exclude the base `where`, exactly as
  // TanStack defines them). The server composes `base AND whereCurrent` (ties, unbounded)
  // and `base AND whereFrom` (next page, bounded by `limit`). The cursor
  // double-read is ONE frame so the server reads both halves at a single `seq`
  // (atomic) and the client applies the page in stream order before any later
  // delta — see ADR-0003. Ties must be unbounded while next is limited, an
  // asymmetry a single (where, limit) can't express, hence two predicates.
  | {
      t: "fetch"
      fetchId: string
      collection: string
      where?: WireExpression
      cursor?: { whereFrom: WireExpression; whereCurrent: WireExpression }
      orderBy?: WireOrderBy
      limit?: number
    }

export type ServerFrame =
  // Snapshot rows (full row) then a boundary; client truncates + applies.
  | { t: "snap"; sub: string; key: unknown; row: unknown; seq: Cursor }
  | { t: "snap-end"; sub: string; seq: Cursor }
  // Live delta; `cols` is a partial (top-level) patch, absent for delete.
  | { t: "d"; sub: string; key: unknown; op: RowOp; cols?: Record<string, unknown>; seq: Cursor }
  // Batch boundary — client commits the buffered sync transaction here.
  | { t: "uptodate"; seq: Cursor }
  // Mutation receipt (the no-subscription-match path lives here; ADR-0002 C1/C2).
  | { t: "committed"; txId: TxId; seq: Cursor; result?: unknown }
  | { t: "rejected"; txId: TxId; error: { code?: string; message: string } }
  // Compaction/rotation reset — client truncates and resnapshots (ADR-0002 C5).
  | { t: "reset"; sub?: string }
  // Response to a `fetch`: the page's rows, in one frame (no live sub).
  | { t: "page"; fetchId: string; rows: Array<unknown>; seq: Cursor }

export type Frame = ClientFrame | ServerFrame
export type FrameTag = Frame["t"]

File: packages/do_sync_client/lib/src/wire/frames.dart
// Wire protocol frames — the Dart mirror of `src/wire/frames.ts` (ADR-0001
// §single-ordered-stream, ADR-0002 §2, ADR-0019 D1).
//
// One ordered stream per DO. The client tracks a single cursor (appliedSeq);
// `seq` is opaque (a stringified bigint). Confirmation rides the same stream:
// `committed`/`rejected` correlate to a client `txId`; there is no second ack
// channel.
//
// Field names and shapes match the TS union exactly; optional fields are
// OMITTED from the wire map when null (the TS decoder reads an absent key as
// `undefined`, the same "not provided" the TS client sends).

/// `insert` | `update` | `delete`.
enum RowOp {
  insert,
  update,
  delete;

  static RowOp fromWire(String s) => switch (s) {
        'insert' => RowOp.insert,
        'update' => RowOp.update,
        'delete' => RowOp.delete,
        _ => throw FormatException('unknown row op: $s'),
      };
}

/// One row operation inside a `mut` frame. `cols` is the full row for an
/// insert, a top-level partial for an update, absent for a delete.
class MutOp {
  const MutOp.insert(this.key, Map<String, Object?> this.cols) : type = RowOp.insert;
  const MutOp.update(this.key, Map<String, Object?> this.cols) : type = RowOp.update;
  const MutOp.delete(this.key) : type = RowOp.delete, cols = null;

  final RowOp type;
  final Object? key;
  final Map<String, Object?>? cols;

  Map<String, Object?> toWire() => {
        'type': type.name,
        'key': key,
        if (cols != null) 'cols': cols,
      };
}

sealed class Frame {
  const Frame();

  Map<String, Object?> toWire();

  /// Decode a wire map into a typed frame. Unknown tags throw [FormatException]
  /// — the transport drops undecodable messages, mirroring the TS client.
  static Frame fromWire(Map<String, Object?> m) {
    final t = m['t'];
    return switch (t) {
      'sub' => SubFrame(
          subId: m['subId'] as String,
          collection: m['collection'] as String,
          where: m['where'],
          orderBy: m['orderBy'],
          limit: (m['limit'] as num?)?.toInt(),
          offset: (m['offset'] as num?)?.toInt(),
          since: m['since'] as String?,
        ),
      'unsub' => UnsubFrame(subId: m['subId'] as String),
      'mut' => MutFrame(
          txId: m['txId'] as String,
          collection: m['collection'] as String,
          ops: (m['ops'] as List<Object?>).map((o) {
            final op = o as Map<String, Object?>;
            final type = RowOp.fromWire(op['type'] as String);
            return switch (type) {
              RowOp.insert => MutOp.insert(op['key'], (op['cols'] as Map<String, Object?>?) ?? const {}),
              RowOp.update => MutOp.update(op['key'], (op['cols'] as Map<String, Object?>?) ?? const {}),
              RowOp.delete => MutOp.delete(op['key']),
            };
          }).toList(),
        ),
      'call' => CallFrame(txId: m['txId'] as String, name: m['name'] as String, args: m['args']),
      'fetch' => FetchFrame(
          fetchId: m['fetchId'] as String,
          collection: m['collection'] as String,
          where: m['where'],
          cursor: m['cursor'] as Map<String, Object?>?,
          orderBy: m['orderBy'],
          limit: (m['limit'] as num?)?.toInt(),
        ),
      'snap' => SnapFrame(sub: m['sub'] as String, key: m['key'], row: m['row'], seq: m['seq'] as String),
      'snap-end' => SnapEndFrame(sub: m['sub'] as String, seq: m['seq'] as String),
      'd' => DeltaFrame(
          sub: m['sub'] as String,
          key: m['key'],
          op: RowOp.fromWire(m['op'] as String),
          cols: m['cols'] as Map<String, Object?>?,
          seq: m['seq'] as String,
        ),
      'uptodate' => UptodateFrame(seq: m['seq'] as String),
      'committed' => CommittedFrame(txId: m['txId'] as String, seq: m['seq'] as String, result: m['result']),
      'rejected' => RejectedFrame(
          txId: m['txId'] as String,
          code: (m['error'] as Map<String, Object?>?)?['code'] as String?,
          message: ((m['error'] as Map<String, Object?>?)?['message'] as String?) ?? 'rejected',
        ),
      'reset' => ResetFrame(sub: m['sub'] as String?),
      'page' => PageFrame(
          fetchId: m['fetchId'] as String,
          rows: m['rows'] as List<Object?>,
          seq: m['seq'] as String,
        ),
      _ => throw FormatException('unknown frame tag: $t'),
    };
  }
}

// --- Client -> server -------------------------------------------------------

class SubFrame extends Frame {
  const SubFrame({
    required this.subId,
    required this.collection,
    this.where,
    this.orderBy,
    this.limit,
    this.offset,
    this.since,
  });
  final String subId;
  final String collection;
  final Object? where;
  final Object? orderBy;
  final int? limit;
  final int? offset;
  final String? since;

  @override
  Map<String, Object?> toWire() => {
        't': 'sub',
        'subId': subId,
        'collection': collection,
        if (where != null) 'where': where,
        if (orderBy != null) 'orderBy': orderBy,
        if (limit != null) 'limit': limit,
        if (offset != null) 'offset': offset,
        if (since != null) 'since': since,
      };
}

class UnsubFrame extends Frame {
  const UnsubFrame({required this.subId});
  final String subId;

  @override
  Map<String, Object?> toWire() => {'t': 'unsub', 'subId': subId};
}

class MutFrame extends Frame {
  const MutFrame({required this.txId, required this.collection, required this.ops});
  final String txId;
  final String collection;
  final List<MutOp> ops;

  @override
  Map<String, Object?> toWire() => {
        't': 'mut',
        'txId': txId,
        'collection': collection,
        'ops': ops.map((o) => o.toWire()).toList(),
      };
}

class CallFrame extends Frame {
  const CallFrame({required this.txId, required this.name, this.args});
  final String txId;
  final String name;
  final Object? args;

  @override
  // `args` is always present (the TS client sends `undefined` args as an
  // explicit member; the server reads both null and absent as "no args").
  Map<String, Object?> toWire() => {'t': 'call', 'txId': txId, 'name': name, 'args': args};
}

class FetchFrame extends Frame {
  const FetchFrame({
    required this.fetchId,
    required this.collection,
    this.where,
    this.cursor,
    this.orderBy,
    this.limit,
  });
  final String fetchId;
  final String collection;
  final Object? where;
  final Map<String, Object?>? cursor; // { whereFrom, whereCurrent }
  final Object? orderBy;
  final int? limit;

  @override
  Map<String, Object?> toWire() => {
        't': 'fetch',
        'fetchId': fetchId,
        'collection': collection,
        if (where != null) 'where': where,
        if (cursor != null) 'cursor': cursor,
        if (orderBy != null) 'orderBy': orderBy,
        if (limit != null) 'limit': limit,
      };
}

// --- Server -> client -------------------------------------------------------

class SnapFrame extends Frame {
  const SnapFrame({required this.sub, required this.key, required this.row, required this.seq});
  final String sub;
  final Object? key;
  final Object? row;
  final String seq;

  @override
  Map<String, Object?> toWire() => {'t': 'snap', 'sub': sub, 'key': key, 'row': row, 'seq': seq};
}

class SnapEndFrame extends Frame {
  const SnapEndFrame({required this.sub, required this.seq});
  final String sub;
  final String seq;

  @override
  Map<String, Object?> toWire() => {'t': 'snap-end', 'sub': sub, 'seq': seq};
}

class DeltaFrame extends Frame {
  const DeltaFrame({required this.sub, required this.key, required this.op, this.cols, required this.seq});
  final String sub;
  final Object? key;
  final RowOp op;
  final Map<String, Object?>? cols;
  final String seq;

  @override
  Map<String, Object?> toWire() => {
        't': 'd',
        'sub': sub,
        'key': key,
        'op': op.name,
        if (cols != null) 'cols': cols,
        'seq': seq,
      };
}

class UptodateFrame extends Frame {
  const UptodateFrame({required this.seq});
  final String seq;

  @override
  Map<String, Object?> toWire() => {'t': 'uptodate', 'seq': seq};
}

class CommittedFrame extends Frame {
  const CommittedFrame({required this.txId, required this.seq, this.result});
  final String txId;
  final String seq;
  final Object? result;

  @override
  Map<String, Object?> toWire() => {
        't': 'committed',
        'txId': txId,
        'seq': seq,
        if (result != null) 'result': result,
      };
}

class RejectedFrame extends Frame {
  const RejectedFrame({required this.txId, this.code, required this.message});
  final String txId;
  final String? code;
  final String message;

  @override
  Map<String, Object?> toWire() => {
        't': 'rejected',
        'txId': txId,
        'error': {if (code != null) 'code': code, 'message': message},
      };
}

class ResetFrame extends Frame {
  const ResetFrame({this.sub});
  final String? sub;

  @override
  Map<String, Object?> toWire() => {'t': 'reset', if (sub != null) 'sub': sub};
}

class PageFrame extends Frame {
  const PageFrame({required this.fetchId, required this.rows, required this.seq});
  final String fetchId;
  final List<Object?> rows;
  final String seq;

  @override
  Map<String, Object?> toWire() => {'t': 'page', 'fetchId': fetchId, 'rows': rows, 'seq': seq};
}

/// A decoded row is always a string-keyed map on this wire.
typedef Row = Map<String, Object?>;
